cmd/kubeadm/app/discovery/discovery.go

```diff
@@ -20,14 +20,68 @@ import (
 	"fmt"
 	"io/ioutil"
 	"net/http"
+	"net/url"
+	"strings"
 
 	"k8s.io/client-go/tools/clientcmd"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/discovery/file"
+	"k8s.io/kubernetes/cmd/kubeadm/app/discovery/https"
+	"k8s.io/kubernetes/cmd/kubeadm/app/discovery/token"
 	kubenode "k8s.io/kubernetes/cmd/kubeadm/app/node"
 	tokenutil "k8s.io/kubernetes/cmd/kubeadm/app/util/token"
 )
 
+func Check(discURL, discFile, discToken string) (string, error) {
+	count := 0
+	disc := ""
+	if discURL != "" {
+		disc = discURL
+		count++
+	}
+	if discFile != "" {
+		disc = discFile
+		count++
+	}
+	if discToken != "" {
+		disc = discToken
+		count++
+	}
+	if count > 1 {
+		return "", fmt.Errorf("too many discovery options chosen: count %d", count)
+	}
+
+	return disc, nil
+}
+
+func Assign(d *kubeadmapi.Discovery, disc string) error {
+	u, err := url.Parse(disc)
```

Comment:
we should get rid of the internal usage of URLs now when we've voted against it in general